keps/sig-api-machinery/20190228-priority-and-fairness.md

```diff
@@ -0,0 +1,370 @@
+  criteria.  For example, as in the CoDel technique --- which will
+  reject a request at the head of the queue if the latency is found to
+  be too big at certain times.
+
+- Based on some resource limit (e.g., QPS or concurrency) and with
+  regards to priority and fairness criteria, requests are dispatched
+  from the queues to be served (i.e., continue down the handler
+  chain).
+
+- We assume that when the requst-timeout handler aborts a request it
+  is effective --- we assume the request stops consuming CPU and
+  memory at that point.
+
+One of the biggest questions is how to formulate the scheduling
+parameters.  There are several related concepts in the state of the
+art of scheduling, and we are trying to figure out what to adopt
+and/or invent.  We would prefer to invent as little as possible; it is
+a non-trivial thing to invent a new --- and sufficiently efficient ---
+scheduling algorithm and prove it correct.  The vmware product line
+uses scheduling parameters named "reservation", "limit", and "shares"
+for each class of workload.  The first two are known elsewhere as
+"assured" and "ceiling".  Finding a published algorithm that
+implements all three has not proven easy.  Alternatively, priorities
+are easy to implement and arguably more desirable --- provided there
+is some form of fairness within each priority level.  The current
+thinking is in that direction: use priorities, with simple equal
+fairness among some categories of requests in each priority level.
+There are published scheduling algorithms that provide fairness, and
+we hope to use/adapt one of them to apply independently within the
+confines of each priroity level.
+
+Another issue is whether to manage QPS or concurrency or what.
+Managing QPS leaps first to mind, perhaps because it is a simple
+concept and perhaps because it is familiar from the self-restraint
+that clients apply today.  But we want to also take service time into
+account; a request flow with longer service times should get less QPS
+because its requests are "heavier" --- they impose more load on the
+apiserver.  A natural way to do this is with an inverse linear
+relation.  For example, when two CPU-bound request flows are getting
+equal CPU from the apiserver, and the first flow's requests have a
+service time that is X times the service time of the second flow's
+requests, the first flow's QPS is 1/X of the second's.  This is
+exactly analogous to what happens in networking: if two flows are
+getting the same bandwidth, and one flow's packets are X times as big
+as the second's, then the first flow's packets per second rate is 1/X
+that of the second flow.  This inverse linear relation amounts to
+managing the product of QPS * service time.  That is equivalent to
+managing concurrency.  Managing concurrency is an obvious choice for
+memory, and we now see it is a good choice for CPU too.  This is also
+a convenient choice because it is what the max-in-flight handler is
```

Comment:
This section made me think of the max-in-flight style limit much more favorably. I will modify my doc to incorporate such a limit when I get a chance.